File: backend/app/api/repository.py
import traceback
from chromadb import PersistentClient
from fastapi import APIRouter, HTTPException

from app.schemas.repository import (
    RepositoryImportRequest,
    RepositoryImportResponse,
)
from app.services.analysis_service import AnalysisService
from app.services.repository_service import RepositoryService

router = APIRouter(tags=["Repositories"])

repository_service = RepositoryService()
analysis_service = AnalysisService()


@router.post(
    "/repositories/import",
    response_model=RepositoryImportResponse,
)
async def import_repository(request: RepositoryImportRequest):
    try:
        repo = repository_service.clone(
            request.github_url
        )

        return RepositoryImportResponse(
            message="Repository cloned successfully.",
            repository_name=repo,
        )

    except Exception:
        print("=" * 80)
        print("IMPORT ENDPOINT FAILED")
        traceback.print_exc()
        print("=" * 80)
        raise


@router.get("/repositories")
def list_repositories():
    client = PersistentClient(
        path="./storage/chromadb"
    )

    return sorted(
        [
            collection.name
            for collection in client.list_collections()
        ]
    )


@router.get("/repositories/{repository_name}/map")
def repository_map(repository_name: str):
    try:
        return analysis_service.generate_repository_map(
                repository_name
            )

    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=str(e),
        )


@router.delete("/repositories/{repository_name}")
def delete_repository(repository_name: str):
    try:
        return repository_service.delete(
            repository_name
        )

    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=str(e),
        )

File: frontend/src/api/repository.js
import axios from "axios";

const api = axios.create({
  baseURL: "http://127.0.0.1:8000/api/v1",
});

export async function getRepositories() {
  const response = await api.get("/repositories");
  return response.data;
}

export async function importRepository(
  githubUrl
) {
  const response = await api.post(
    "/repositories/import",
    {
      github_url: githubUrl,
    }
  );

  return response.data;
}

export async function getRepositoryMap(
  repository
) {
  const response = await api.get(
    `/repositories/${repository}/map`
  );

  return response.data;
}

export async function deleteRepository(
  repository
) {
  await api.delete(
    `/repositories/${repository}`
  );
}